### ISSUE TITLE
RFC: Use conditional compilation to make sf_error an optional include for special function compiled kernels.

### ISSUE THREAD

**Issue by steppi** [scipy/scipy#19480]
Following up on https://github.com/scipy/scipy/issues/19404, we are moving forward with preliminary work that could lead to making an important subset of special a header only library of NVCC compatible code that could be included in other projects (and which may eventually find a home outside of the SciPy code base.

As discussed in the review comments for https://github.com/scipy/scipy/pull/19435, we use code from `sf_error.c` for error handling in special functions, which makes calls to the Python C API. While this is the case, any code which makes use of `sf_error` cannot be NVCC compatible. One solution is to use conditional compilation to make make the use of `sf_error` only occur when a compile time symbol such as `USE_SF_ERROR` is set. There are several ways this could be done, so I'm creating this RFC issue for discussion on how this could best be achieved.

One possibility would be to define a new header file with the following lines; defining `sf_error` to be a no-op if the symbol `USE_SF_ERROR` is not defined. 

```C++
#ifdef USE_SF_ERROR
#include "sf_error.h"
#else
inline void sf_error(const char *func_name, sf_error_t code, const char *fmt, ...) {};
#endif
```

For any compiled kernel written in C++ where we want to make the use of `sf_error` optional , we could include this header instead of `"sf_error.h"`. I think this would be would be simpler and less intrusive then modifying the current `sf_error` code directly, with the cost that we'd have to remember to include the right header instead of having things work seamlessly.

I'm happy to hear your thoughts everyone.

**Comment by steppi** [comment]
Thanks @izaid, I like that idea. I think I misunderstand something (sincerely)

In the following lines, it seems like `SP_SPECFUN_NOERROR` should correspond to the case where `set_error` does nothing. At least that's how I understand `NOERROR`.

```C++
#ifndef SP_SPECFUN_NOERROR
        void set_error(const char *func_name, int code, const char *fmt, ...) {
            // nothing
        }
#else
        void set_error(const char *func_name, int code, const char *fmt, ...);
#endif

```

In any case, I'll put together a PR implementing your idea. Plenty of time to bikeshed stuff like names.

**Comment by izaid** [comment]
@steppi Yes, I get you, probably `SP_SPECFUN_NOERROR` should be renamed `SP_SPECFUN_ERROR`. I think the important thing is that error is opt in, thus a library has to define something to not get the default empty implementation.

**Comment by steppi** [comment]
@izaid, this solution won't work for Cephes if we make it a header only C library, since it uses `sf_error` internally as well. Do you think it's fine to go full steam ahead with this approach for now, and sort Cephes out later, or should we rethink things in the context of Cephes?

**Comment by izaid** [comment]
@steppi Good find! I guess cephes didn't originally use `sf_error`, and that SciPy at some point just added it into cephes.

In that case, I think the cephes functions would need to move from `sf_error` to `scipy::special::set_error` as well.

I'm fine to leave cephes alone for now and go with this approach. We'll have to get cephes to hook into this at some point, but it doesn't need to be now. There are options.
